lottery/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+#![no_std]
+
+#[cfg(test)]
+mod simple_tests;
+
+#[cfg(test)]
+mod panic_tests;
+
+use codec::{Decode, Encode};
+use gstd::{debug, exec, msg, prelude::*, ActorId};
+use lt_io::*;
+use scale_info::TypeInfo;
+use sp_core::hashing::blake2_256;
+
+const ZERO_ID: ActorId = ActorId::new([0u8; 32]);
+
+#[derive(Debug, Default, Encode, Decode, TypeInfo)]
+struct Lottery {
+    lottery_state: LotteryState,
+    lottery_owner: ActorId,
+    players: BTreeMap<u32, Player>,
+    players_timestamp: BTreeMap<ActorId, u64>,
+    lottery_history: BTreeMap<u32, ActorId>,
+    lottery_id: u32,
+}
+
+impl Lottery {
+    // checks that lottery has started and lottery time has not expired
+    fn lottery_is_on(&mut self) -> bool {
+        self.lottery_state.lottery_started
+            && (self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration)
+                > exec::block_timestamp()
+    }
+
+    /// Launches a lottery
+    /// Requirements:
+    /// * Only owner can launch lottery
+    /// * Lottery must not have been launched earlier
+    /// Arguments:
+    /// * `duration`: lottery duration in milliseconds
+    fn start_lottery(&mut self, duration: u64) {
+        if msg::source() == self.lottery_owner && !self.lottery_is_on() {
+            self.lottery_state.lottery_started = true;
+            self.lottery_state.lottery_start_time = exec::block_timestamp();
+            self.lottery_state.lottery_duration = duration;
+            self.lottery_id += 1;
+        } else {
+            panic!(
+                "start_lottery(): Lottery on: {}  Owner message: {}",
+                self.lottery_is_on(),
+                msg::source() == self.lottery_owner
+            );
+        }
+    }
+
+    /// Adding a lottery player
+    /// Requirements:
+    /// * Lottery has started and lottery time has not expired
+    /// * Contribution must be greater than zero
+    /// * Player cannot be added again
+    fn enter(&mut self) {
+        if self.lottery_is_on() && msg::value() > 0 {
+            if let collections::btree_map::Entry::Vacant(new_player) =
```

Comment:
Too long path. Use imports and Rust idiom for `use`.

https://doc.rust-lang.org/stable/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html

```suggestion
            if let Entry::Vacant(new_player) =
```

Reply:
Now it looks like this:
if self
    .players
    .values()
    .find(|player| player.player_id == msg::source())
    .is_none()